### PR TITLE
code change for  extensible design

### PR DESCRIPTION
**Pull request by Sachin4403**

### CODE REVIEW

**Comment by joe94** [comment]
This looks good from the standpoint of #12. I linked the issue to be closed with this PR. I'll let my teammates comment on the other issues.

File: config.go
Comment (kristinaspring) [inline config.go:21]: This isn't a sentence.  It would also be nice to provide a bit more information.

File: config.go
Comment (kristinaspring) [inline config.go:29]:
```suggestion
	// Endpoint is the endpoint to which spans need to be submitted.
```

File: config.go
Comment (kristinaspring) [inline config.go:36]: This should be a sentence and start with "Providers".

File: config.go
Comment (kristinaspring) [inline config.go:39]: This should be a sentence and start with "SpanIDHeaderName".

File: config.go
Comment (kristinaspring) [inline config.go:41]: Do we want to add a comment for `TraceIDHeaderName`?

File: config.go
Comment (kristinaspring) [inline config.go:44]: Instead of describing what TraceConfig contains, can you provide information on what it does and how a consumer would want to use it?

File: config.go
Comment (kristinaspring) [inline config.go:50]:
```suggestion
// NewTraceConfig create a new TraceConfig object.
```

This is how I would change the comment to follow Effective Go guidelines (starts with "NewTraceConfig" and is a complete sentence).  But I'm not sure we need this function? Is there a reason why we can't just export `config` and `traceProvider` in the `TraceConfig` struct?

File: config.go
Comment (kristinaspring) [inline config.go:45]: This seems to be a struct specifically for use in the middleware, is that true? I don't think the middleware needs all of the configuration values provided in the `Config` struct - it looks like it only needs `SpanIDHeaderName` and `TraceIDHeaderName`.  I think it would make more sense for the TraceConfig to only have access to those values, instead of many values it doesn't need.

It looks like`ConfigureTracerProvider()` doesn't need these two header names. We can split the `Config` struct into two separate structs for these two separate purposes.

File: middleware.go
Comment (kristinaspring) [inline middleware.go:33]: This comment doesn't follow Effective Go guidelines. Please see other comments for examples on how it can be fixed, or let me know if you want further guidance.

File: tracer.go
Comment (kristinaspring) [inline tracer.go:27]: This comment doesn't follow Effective Go guidelines. Here is an example of how I would fix it:
```suggestion
// InjectTraceInformationInLogger adds the trace id and span id to key value pairs that can 
// be provided to a logger.
```

File: tracer.go
Comment (kristinaspring) [inline tracer.go:39]: This comment doesn't follow Effective Go guidelines. Please see other comments for examples on how it can be fixed, or let me know if you want further guidance.

File: tracer.go
Comment (kristinaspring) [inline tracer.go:49]: This comment doesn't follow Effective Go guidelines. Please see other comments for examples on how it can be fixed, or let me know if you want further guidance.

File: tracer.go
Comment (kristinaspring) [inline tracer.go:57]: This comment doesn't follow Effective Go guidelines. Please see other comments for examples on how it can be fixed, or let me know if you want further guidance.

**Review by kristinaspring — CHANGES_REQUESTED** [review]
I tried to focus on my issues as well.  Returning the `TracerProvider` looks good! The comments still need some work.

File: candlelight.go
Comment (kcajmagic) [inline candlelight.go:74]: The `flushFn` should be called when the application is closing. This allows jaeger to export the remaining spans.
https://pkg.go.dev/go.opentelemetry.io/otel/exporters/trace/jaeger#Exporter.Flush

It might be easier to ignore the `flushFn` instead of returning it for later usage.

**Review by kcajmagic — CHANGES_REQUESTED** [review]

**Review by joe94 — COMMENTED** [review]
Just FYI I don't have additional comments to those that my teammates had. @Sachin4403, in case you were waiting for everybody to review the PR before starting on the changes, feel free to start as we can always keep adding additional comments later. Thanks!

File: candlelight.go
Comment (Sachin4403) [reply candlelight.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: candlelight.go
Comment (Sachin4403) [reply candlelight.go:60]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go:49]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Removed the NewTraceConfig function and made config and traceProvider exportable.

File: config.go
Comment (Sachin4403) [reply config.go:54]: Created a separate config for Header

File: middleware.go
Comment (Sachin4403) [reply middleware.go]: Modified comments as suggested.

File: tracer.go
Comment (Sachin4403) [reply tracer.go]: Modified comments as suggested.

File: tracer.go
Comment (Sachin4403) [reply tracer.go]: Modified comments as suggested.

File: tracer.go
Comment (Sachin4403) [reply tracer.go]: Modified comments as suggested.

File: candlelight.go
Comment (Sachin4403) [reply candlelight.go]: Removed the `flushfn`

File: tracer.go
Comment (Sachin4403) [reply tracer.go]: Modified comments as suggested.

File: config.go
Comment (Sachin4403) [reply config.go]: Modified comments as suggested.

**Review by kristinaspring — APPROVED** [review]
This looks awesome! 🎉

**Review by kcajmagic — APPROVED** [review]
🥳  Looks good

**Review by joe94 — APPROVED** [review]
Linked up the issues so they auto-close. This looks great! 🚀

**Review by johnabass — APPROVED** [review]
Looks great!

**Pull request merged**
